docker/configure_workers_and_start.py

```diff
@@ -427,108 +782,101 @@ def generate_worker_files(
             listeners += original_listeners
 
     # The shared homeserver config. The contents of which will be inserted into the
-    # base shared worker jinja2 template.
-    #
-    # This config file will be passed to all workers, included Synapse's main process.
+    # base shared worker jinja2 template. This config file will be passed to all
+    # workers, included Synapse's main process. It is intended mainly for disabling
+    # functionality when certain workers are spun up, and adding a replication listener.
     shared_config: Dict[str, Any] = {"listeners": listeners}
 
     # List of dicts that describe workers.
     # We pass this to the Supervisor template later to generate the appropriate
     # program blocks.
     worker_descriptors: List[Dict[str, Any]] = []
 
-    # Upstreams for load-balancing purposes. This dict takes the form of a worker type to the
-    # ports of each worker. For example:
+    # Upstreams for load-balancing purposes. This dict takes the form of the base worker
+    # name to the ports of each worker. For example:
```

Comment:
The `worker_base_name` is going to be either something custom(like poor `bob` that I keep picking on) or it will be the deterministic set of `worker_types`. Since the definition is broader, it made since to update the comment to reflect that. 

Specifically, this allows for the capability for something like: `WORKER_TYPES='user_dir+event_creator, user_dir+federation_inbound'` to correctly load-balance the `user_dir` endpoints to either worker

Reply:
Why not define upstreams to be the same as worker roles/types, then make the worker a member of each upstream worker type it can handle.

E.g, `bob=user_dir+event_creator, kat=event_creator+pusher`

⇒

```
upstream user_dir {
    bob
}

upstream event_creator {
    bob kat
}

upstream pusher {
    kat
}
```

(imagine that was nginx syntax)

Reply:
Well, what happens right now(for your example):
```
upstream bob__kat {
    server localhost:18010;
    server localhost:18011;
}
```
and then later on, the endpoints:
```
location ~* ^/_matrix/federation/(v1|v2)/user/devices/ {
    proxy_pass http://localhost:18010;
    ...
}

location ~* ^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/(join|invite|leave|ban|unban|kick)$ {
    proxy_pass http://bob__kat;
    ...
}
```
And there are other endpoints similar to the second, but with different port numbers(depending on worker directed at). Since the point of upstreams(in this context, Complement-centric) is being able to serve a given endpoint over any of the ports allowed, if there is only a single port then the upstream block is pointless and basically an extra 3 lines in a file that would be extra middle man.

All that being said, I will ponder on this for a while. I think I can see where you are going but don't currently have an argument against it.

Reply:
> I'm not sure why we'd switch over from type to base name here — what's wrong with just using types?

🤦‍♂️ TL;DR: because this is a beneficial and flexible side effect of being able to name a worker while also merging functionality. Have to call it something and sometimes that something is shorter than 8 worker_types concatenated with a `+`, just don't want that something to be potentially pointing at something else. Hence, unless we have requested a specific name, it *is* worker_types. But they are sorted alphabetically and concatenated with a `+`, which represents a given, specific, worker. Otherwise, the requested worker name is only used in the docker logs as a visible place.

(Also, I'm not sure what the line length is for an upstream name in nginx, I couldn't find a concrete reference. It is written in C, though, which if it's compiled with GCC means it has a potential of 100,000 chars, but will complain after 510. We better never come close to that)

Reply:
I'm actually perfectly happy to have duplicate, or single-entry, upstreams if we name the upstreams after the worker roles rather than the combination of the worker names.

I think this would still be clear to a reader — in a sense, it would also help to give a listing of all the workers that handle certain classes of endpoints.